src/uct/ib/ud/accel/ud_mlx5.c

```diff
@@ -176,79 +189,149 @@ static UCS_CLASS_DEFINE_NEW_FUNC(uct_ud_mlx5_ep_t, uct_ep_t, uct_iface_h,
 UCS_CLASS_DEFINE_DELETE_FUNC(uct_ud_mlx5_ep_t, uct_ep_t);
 
 
-static ucs_status_t
-uct_ud_mlx5_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t hdr,
-                        const void *buffer, unsigned length)
+/*
+ * Generic inline+iov post-send function
+ * The caller should check that header size + sg list would not exceed WQE size.
+ */
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_ud_mlx5_ep_inline_iov_post(uct_ep_h tl_ep, uint8_t am_id,
+                               /* inl. header */ const void *header, size_t header_size,
+                               /* inl. data */   const void *data, size_t data_size,
+                               /* iov data */    const uct_iov_t *iov, size_t iovcnt,
+                               uint32_t packet_flags, uct_completion_t *comp,
+                               unsigned stat_ops_counter, unsigned stat_bytes_counter,
+                               const char *func_name)
 {
-    uct_ud_mlx5_ep_t *ep = ucs_derived_of(tl_ep, uct_ud_mlx5_ep_t);
     uct_ud_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface,
                                                 uct_ud_mlx5_iface_t);
-    size_t ctrl_av_size = uct_ud_mlx5_ep_ctrl_av_size(ep);
-    struct mlx5_wqe_ctrl_seg *ctrl;
+    uct_ud_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_ud_mlx5_ep_t);
     struct mlx5_wqe_inl_data_seg *inl;
-    uct_ud_am_short_hdr_t *am;
-    uct_ud_neth_t *neth;
+    struct mlx5_wqe_ctrl_seg *ctrl;
+    size_t inline_size, wqe_size;
     uct_ud_send_skb_t *skb;
-    size_t wqe_size;
+    ucs_status_t status;
+    uct_ud_neth_t *neth;
+    void *wqe_data;
 
-    /* data a written directly into tx wqe, so it is impossible to use
-     * common ud am code
-     */
-    UCT_CHECK_AM_ID(id);
-    UCT_CHECK_LENGTH(sizeof(uct_ud_neth_t) + sizeof(hdr) + length,
-                    0, iface->super.config.max_inline, "am_short");
+    UCT_CHECK_AM_ID(am_id);
+    UCT_UD_CHECK_ZCOPY_LENGTH(&iface->super, header_size + data_size,
+                              uct_iov_total_length(iov, iovcnt));
+    UCT_CHECK_IOV_SIZE(iovcnt, uct_ud_mlx5_max_am_iov(), func_name);
 
     uct_ud_enter(&iface->super);
 
     skb = uct_ud_ep_get_tx_skb(&iface->super, &ep->super);
     if (!skb) {
-        uct_ud_leave(&iface->super);
-        return UCS_ERR_NO_RESOURCE;
+        status = UCS_ERR_NO_RESOURCE;
+        goto out;
+    }
+
+    wqe_size        = uct_ud_mlx5_ep_get_next_wqe(iface, ep, &ctrl, (void**)&inl);
```

Comment:
the initialization can be reorganized as follows:
```
inline_size     = sizeof(*neth) + header_size + data_size;
wqe_size        = uct_ud_mlx5_ep_get_next_wqe(iface, ep, &ctrl, (void**)&inl) +
                  sizeof(*inl) + inline_size;
inl->byte_count = htonl(inline_size | MLX5_INLINE_SEG);
skb->len        = inline_size;
```

Reply:
since IMO uct_ud_mlx5_ep_get_next_wqe() should not return wqe size, it's not possible (see next comment)